### ISSUE TITLE
Firebase Auth plugin: currentUser only from memory

### ISSUE THREAD

**Issue by mkhennoussi** [flutter/flutter#10624]
Hello ! :) Thank you all for Flutter and all the plugins :) !

I am not 100% sure, but once the user is logged in, the Firebase SDK store the currentUser on the device (in SharedPrefs or something). It is useful if we want to show the login screen or the home screen, depending if the user is logged in or not.

The current version of the Firebase Auth Plugin doesn't seem to get the currentUser from the device, but only from memory (it stores the logged user in a variable: `FirebaseUser _currentUser;`  ). So when the app restarts, we loose the currentUser variable and the user has to log in again.

Is there any plan in the future to have a method to get the currentUser stored on the device ?

**Comment by mit-mit** [comment]
cc @collinjackson

**Comment by branflake2267** [comment]
You won't want to persist the current user, because you don't know how long that object could be valid for. Given that the oauth token could expire. Instead use the sdk to figure if the current user is active and logged in. This can be done with the google_sign_in plugin and firebase_auth plugin. 

For example, when the pages state resumes, you'll want to restart the entire sing in process. Let's say the user pauses to check email, in general, the page will be paused and cached. Then let's say I come back a day later, and for some reason, I should sign in again. You won't ever know if the user login state is being audited, expiring, or revoked. So giving this, you'll want to always let the api do this heavy lifting, by requesting the current user, and if it doesn't exist, then call the sign in process. Most often the device sign in with out input, but sometimes it requires more steps, handled by the native device routine. 

If you do want to persist something for the firebase user, you could persist the uid for comparison and use in calls to the database. 

Example of the call on the pages init state:
https://github.com/flutter/plugins/blob/master/packages/google_sign_in/example/lib/main.dart#L38

**Comment by mkhennoussi** [comment]
@branflake2267 Thank you very much for your answer. Sorry for the confusion,  I didn't say I wanted to persist the current user, I just say that I thought the SDK had some kind of persistency, because with the native approach, if the app is paused/stopped then resumed, we can still access the current user (even if the token is not valid anymore). I agree with you about the token, on the native approach, I use the AuthStateListener to avoid theses cases. 
I am using only the email & password method, so I don't know yet if the google_sign_in plugin will be useful to figure if the current user is active and logged in. I still don't know yet how to handle the login flow properly without asking the user to log in every time the app is paused or stopped.

I guess they are several good reasons for the Firebase team to persist the current user. They are using it almost everywhere in the docs. Anyway if you think we should not use the same approach, we can close this issue.

**Comment by briblanch** [comment]
@mkhennoussi I had to do exactly what you explained in one of my apps. I have added the functionality you described [here for the dart side](https://github.com/briblanch/plugins/blob/master/packages/firebase_auth/lib/firebase_auth.dart#L146) and [here for iOS](https://github.com/briblanch/plugins/blob/master/packages/firebase_auth/ios/Classes/FirebaseAuthPlugin.m#L92)

This is just kind of hacked together and I have not implemented the Android side, but what I am doing is showing a loading screen while awaiting the `cachedCurrentUser` call.  Then, I show a login screen or another screen based on if the `currentUser` property is set after the `cachedCurrentUser` future completes.

**Comment by collinjackson** [comment]
I agree that we should provide an API that returns a `Future<FirebaseUser>`, like `signInSilently()` in the Google Sign In plugin. We may also want to deprecate synchronous access to `currentUser` to discourage race conditions, see discussion here: https://github.com/flutter/plugins/pull/94

**Comment by branflake2267** [comment]
Speaking of abstracting away some of the logic, this api makes it easier to wire up firebase auth on android, I think this is the best method I've seen so far for wiring it up. Besides, you can also use the other 3 options to login as well. :)

https://github.com/firebase/FirebaseUI-Android/tree/master/auth#using-firebaseui-for-authentication

```
FirebaseAuth auth = FirebaseAuth.getInstance();
if (auth.getCurrentUser() != null) {
    // already signed in
} else {
    // not signed in
}
```

**Comment by sethladd** [comment]
Closing this as a dupe of https://github.com/flutter/flutter/issues/11541 but please correct me if I'm wrong.

**Comment by collinjackson** [comment]
I believe that this was fixed by flutter/plugins#180

**Comment by kroikie** [comment]
@mkhennoussi

This issue has been moved to https://github.com/FirebaseExtended/flutterfire/issues/999. Any further collaboration will be done there.
